### DISCUSSION TITLE
String Interpolation Highlighting

### DISCUSSION THREAD

**Discussion by nathanrboyer** [FedeClaudi/Term.jl#87] [Ideas]
My typical usage of `println` would look something like: `println("These are the results: x = $x, y = $y")`.
I wonder if `tprintln` could automatically color interpolated values? I know I can write `tprintln("These are the results: x =", x, "\b, y =", y)` instead, but I find the former code easier to read.

I did see the portion of the documentation that would presumably cover this use case.
> But, we are working on a highlight feature that will be able to parse strings and color their elements correctly. Like this:
> ```julia
> import Term: tprint, highlight
> tprint(highlight("This is a: 1 of type ::Int64 this is a function print"))
> ```
However, I would rather see `highlight` as a keyword option for `tprint` that defaults to `true` than its own function. This default highlighting would be lowest priority and overwritten by any other formatting called for inside `tprint`. That way if you don't want the element highlighting, you could just color it some other way (or take the `t` off `tprint`).

Finally, I would just like to commend the authors for writing some of the best documentation I have seen. Very enjoyable to read!

**Comment by FedeClaudi** [comment]
Thanks for reaching out and for the comment about the docs, I tried making them less tedious than usual documentation. 

I agree, `tprint` should style the output for you, instead of having to do it manually or invoke `highlight`. The `highlighting` functionality is currently not great, but it's one of the things I want to improve when I get a chance. I will also add automatic styling to `tprint`, thanks for the suggestion!

**Comment by FedeClaudi** [comment]
An update, I've improved `highlight` and added it as the default behavior of `tprint` (for string only for now):

```Julia
txt = """
These are numbers: 12311, 2 122.2321  121e-4 1.5e10

These are strings: "aaa" and 
            'dsda33d'

These are operators: + - * % ^ /

These are operators: [green]+[/green] - * % ^  /


This is code: `print(x)` and
    ```
        do α
        do β
    ```


This is an expression: :(x + y)
"""

print("\n\n")
tprint(txt)
```
<img width="332" alt="image" src="https://user-images.githubusercontent.com/17436313/169708295-27bd4329-f4bd-4424-a71a-e4138419bf0b.png">



This is not bad, but the problem is that user provided markup style is ignored at beast or broken in the worst case.
This is because given a string `"[green]+[/green]` `highlight` surrounds `+` with `[red]+[/red]` which being inside of `green` takes precedence.  Preserving the user's markup's style might be very tricky since it could even be defined several lines before `+` appears... 

For this reason it might be best to not have `highlight` work as the default - or perhaps only have it as default if no markup information is in the string to be printed.
What do you think?


Note: user's markup information will work fine for portions of text that don't get highlighted, it will only be lost for highlighted stuff.

**Reply by nathanrboyer** [reply]
> ... or perhaps only have it as default if no markup information is in the string to be printed.

This would be my preference. Then I think you wouldn't have to use the `highlight` keyword argument in most cases. `tprint("x = $x")` would have nice defaults, and `tprint("{red}x = $x"{/red}")` would respect the user formatting and just be red. Once you are customizing colors, you probably want to pick your own anyway.

**Reply by FedeClaudi** [reply]
This is indeed the behavior now:
```julia
using Term


x = 1
print("\n"^3)

print("print:  ")
println("x = $x")

print("tprint:  ")
tprint("x = $x")


print("\nwith style:  ")
tprint("{red}x{/red} = $x")
```

<img width="142" alt="image" src="https://user-images.githubusercontent.com/17436313/174885073-d1a8cabc-8950-4c08-8f16-fba7d1061fe5.png">

**Reply by nathanrboyer** [reply]
I am suggesting that highlighting be automatically turned off in the last case, only coloring the "x" since markup information is present in the string.
I would always want my markup information to take precedence e.g. this should just be green.
![image](https://user-images.githubusercontent.com/65452054/175044531-7525e8db-85cf-4480-ab38-86c258f9971c.png)

**Comment by FedeClaudi** [comment]
Regarding your original question:
```
This is an expression: :(x + y)
These are the results: x = 1, y = 2
```
results in `=1` and `=2` being highlighted - but it won't work if the interpolated variables are e.g. strings or symbols since once they're interpolated they just look like a normal part of the string. 

If you want strings/symbols to work correctly you'll need to add `:` or `'` in the string:
```Julia
x, y = :z, "test"
tprint("These are the results: x = :$x, y = '$y'")
```

Or you can do:
```Julia
tprint("These are the results: x = $(highlight(x)), y = $(highlight(y))")
```
which would work in most cases except for `x,y` being strings
